crates/symbolicator-proguard/src/symbolication.rs

```diff
@@ -0,0 +1,341 @@
+
+impl ProguardService {
+    pub async fn symbolicate_jvm(
+        &self,
+        request: SymbolicateJvmStacktraces,
+    ) -> CompletedJvmSymbolicationResponse {
+        let SymbolicateJvmStacktraces {
+            scope,
+            sources,
+            exceptions,
+            stacktraces,
+            modules,
+            release_package,
+            ..
+        } = request;
+        let mappers = future::join_all(
+            modules
+                .iter()
+                .map(|module| self.download_proguard_file(&sources, &scope, module.uuid)),
+        )
+        .await;
+
+        // TODO: error handling/reporting
+        let mappers: Vec<_> = mappers
+            .iter()
+            .filter_map(|res| match res {
+                Ok(mapper) => Some(mapper.get()),
+                Err(_e) => None,
+            })
+            .collect();
+
+        let mut remapped_exceptions = Vec::with_capacity(exceptions.len());
+
+        for raw_exception in exceptions {
+            remapped_exceptions
+                .push(Self::map_exception(&mappers, &raw_exception).unwrap_or(raw_exception));
+        }
+
+        let mut remapped_stacktraces = Vec::with_capacity(stacktraces.len());
+
+        for raw_stacktrace in stacktraces {
+            let remapped_frames = raw_stacktrace
+                .frames
+                .iter()
+                .flat_map(|frame| {
+                    Self::map_frame(&mappers, frame, release_package.as_deref()).into_iter()
+                })
+                .collect();
+
+            remapped_stacktraces.push(JvmStacktrace {
+                frames: remapped_frames,
+            });
+        }
+
+        CompletedJvmSymbolicationResponse {
+            exceptions: remapped_exceptions,
+            stacktraces: remapped_stacktraces,
+        }
+    }
+
+    fn map_exception(
+        mappers: &[&proguard::ProguardMapper],
+        exception: &JvmException,
+    ) -> Option<JvmException> {
+        if mappers.is_empty() {
+            return None;
+        }
+
+        let key = format!("{}.{}", exception.module, exception.ty);
+
+        let mapped = mappers.iter().find_map(|mapper| mapper.remap_class(&key))?;
+
+        // TOOD: Capture/log error
+        let (new_module, new_ty) = mapped.rsplit_once('.')?;
+
+        Some(JvmException {
+            ty: new_ty.into(),
+            module: new_module.into(),
+        })
+    }
+
+    fn map_frame(
+        mappers: &[&proguard::ProguardMapper],
+        frame: &JvmFrame,
+        release_package: Option<&str>,
+    ) -> Vec<JvmFrame> {
+        let proguard_frame =
+            proguard::StackFrame::new(&frame.class, &frame.method, frame.lineno as usize);
+        let mut mapped_frames = Vec::new();
+
+        // first, try to remap complete frames
+        for mapper in mappers {
+            mapped_frames.clear();
+            mapped_frames.extend(mapper.remap_frame(&proguard_frame));
+
+            if !mapped_frames.is_empty() {
```

Comment:
```suggestion
            if mapped_frames.is_empty() {
                continue;
            }
```

I’m personally a big fan of early return instead of rightward drift